### PR TITLE
kaiax/gov: Replace remaining with gov

### PR DESCRIPTION
**Pull request by ian0371**
## Proposed changes

This PR is a sequel of #162.

## Types of changes

- [ ] Bugfix
- [x] New feature or enhancement
- [ ] Others

## Checklist

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/kaiachain/kaia/blob/main/CONTRIBUTING.md) doc
- [x] I have read the [CLA](https://gist.github.com/kaiachain-dev/bbf65cc330275c057463c4c94ce787a6) and signed by comment `I have read the CLA Document and I hereby sign the CLA` in first time contribute
- [x] Lint and unit tests pass locally with my changes (`$ make test`)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Further comments
<details>
<summary>MixedEngine method replaced by GovModule</summary>

It excludes `governance` directory, as they will be removed.

- [x] EffectiveParams
- [x] CurrentParams
- [x] AddVote
- [x] ValidateVote
- [x] ReadGovernance
- [x] WriteGovernance
- [x] GetEncodedVote
- [x] VerifyGovernance
- [x] GetVoteMapCopy
- [x] GetGovernanceTalliesCopy
- [x] CurrentSetCopy
- [x] PendingChanges
- [x] IdxCache
- [x] IdxCacheFromDb
- [x] NodeAddress
- [x] TotalVotingPower
- [x] MyVotingPower
- [x] SetNodeAddress
- [x] GetGovernanceItemsFromChainConfig
- [x] SetTxPool
- [x] GetTxPool
- [x] BlockChain
- [x] DB
</details>

<details>
<summary>MixedEngine methods that will be completely removed by ValSet module</summary>

These will be removed after introducing ValSet module.

- UpdateParams
- CanWriteGovernanceState
- WriteGovernanceState
- ClearVotes
- WriteGovernanceForNextEpoch
- UpdateCurrentSet
- HandleGovernanceVote
- Votes
- SetTotalVotingPower
- SetMyVotingPower
- SetBlockchain
</details>

<details>
<summary>ChainConfig replaced by GovModule</summary>

- [x] next basefee in worker
- [x] next basefee in txpool
- [x] `node/cn/backend.go:New`: UnitPrice. Removed because it's unused.
- [x] `blockchain/types/derivesha/mux.go:getType`: DeriveShaImpl.
</details>

<details>
<summary>ChainConfig left as-is</summary>

There are cases when the genesis chain config can be used. Either it's immutable, or not essential to use the latest value.

- `blockchain/genesis.go:SetGenesisGovernance`: initial 13 parameters. Not called by `ken init`.
- UnitPrice
    - `accounts/abi/bind/backends.go:SuggestGasPrice`: UnitPrice. The value does not matter. Also, importing GovModule results in import cycle.
    - `datasync/dbsyncer/dbsync.go:syncBlockHeader`: UnitPrice. The value does not matter.
    - `datasync/dbsyncer/dbsync_context.go:syncBlockHeaderContext`: UnitPrice. The value does not matter.
    - `datasync/dbsyncer/dbsync_multi.go:parallelSyncBlockHeader`: UnitPrice. The value does not matter.
- Governance
    - `blockchain/genesis.go:SetupGenesisBlock`: StakingUpdateInterval, ProposerUpdateInterval. The value is immutable.
    - `cmd/utils/nodecmd/chaincmd.go:initGenesis`: StakingUpdateInterval, ProposerUpdateInterval. The value is immutable.
    - `blockchain/genesis.go:ToBlock`: LowerBoundBaseFee. Genesis config must be used.
    - `kaiax/staking/impl/init.go:Init`: StakingUpdateInterval, UseGiniCoeff, MinimumStake. The value is immutable.
    - `node/cn/handler.go:{handleStakingInfoRequestMsg,handleStakingInfoMsg}`: UseGiniCoeff, MinimumStake. The value is immutable.
    - `blockchain/state_transition.go:TransitionDb`: DeferredTxFee. The value is immutable. Also, adding GovModule to the object is difficult.
    - `cmd/utils/nodecmd/chaincmd.go:ValidateGenesisConfig`: ProposerUpdateInterval, StakingUpdateInterval, GovernanceMode, GoverningNode. Genesis config must be used.
- Istanbul
    - `blockchain/blockchain.go:isCommitTrieRequired`: ProposerPolicy. The value is immutable.
    - `blockchain/blockchain.go:setHeadBeyondRoot`: Epoch. The value is immutable.
    - `consensus/istanbul/backend/engine.go:Finalize`: ProposerPolicy. The value is immutable.
    - `node/cn/handler.go:{handleStakingInfoRequestMsg,handleStakingInfoMsg}`: ProposerPolicy. The value is immutable.
</details>

<details>
<summary>ChainConfig replaced by GovModule</summary>

- [x] next basefee in worker
- [x] next basefee in txpool
- [x] `node/cn/backend.go:New`: UnitPrice. Removed because it's unused.
- [x] `blockchain/types/derivesha/mux.go:getType`: DeriveShaImpl.
</details>

<details>
<summary>ChainConfig that will be completely removed by ValSet module</summary>
These will be removed after introducing ValSet module.

- `consensus/istanbul/backend/backend.go:{ParentValidators,getValidators}`: ProposerPolicy, SubGroupSize.
- `consensus/istanbul/validator/weighted.go:RefreshValSet`: UseGiniCoeff.
</details>

### CODE REVIEW

File: kaiax/gov/headergov/impl/init.go [hunk 1]
```diff
@@ -259,3 +257,71 @@ func validateOpts(opts *InitOpts) error {
...
+	if err != nil {
+		panic(err)
+	}
+
+	return ret
+}
+
+func getGenesisParamNames(config *params.ChainConfig) []gov.ParamName {
+	genesisParamNames := make([]gov.ParamName, 0)
+
```
Comment (blukat29) [inline kaiax/gov/headergov/impl/init.go:290]: - Please comment something like "Note that DeriveShaImpl is not included in genesisBlock.Governance for backwards compatibility"
- In old code https://github.com/klaytn/klaytn/blob/848b00c4a05222c4f74ade60e9ddbe0b9595db2a/blockchain/genesis.go#L464 the UnitPrice field was not in `config.Genesis != nil` condition. The behavior was added later https://github.com/klaytn/klaytn/blob/848b00c4a05222c4f74ade60e9ddbe0b9595db2a/governance/default.go#L1090 and the behavior seems a peculiar bug?

File: kaiax/gov/headergov/impl/init.go [hunk 2]
```diff
@@ -259,3 +257,71 @@ func validateOpts(opts *InitOpts) error {
...
+	govData := headergov.NewGovData(partialParamSet)
+	ret, err := govData.ToGovBytes()
+	if err != nil {
+		panic(err)
+	}
+
+	return ret
+}
+
+func getGenesisParamNames(config *params.ChainConfig) []gov.ParamName {
```
Comment (blukat29) [inline kaiax/gov/headergov/impl/init.go:288]:
```suggestion
// getGenesisParamNames returns a subset of parameters to be included in the genesis block header's Governance field. The subset depends on genesis block's hardfork level and genesis chainconfig. For backward compatibility, some fields may be intentionally omitted.
func getGenesisParamNames(config *params.ChainConfig) []gov.ParamName {
```

File: kaiax/gov/headergov/impl/init.go (hunk 1 above)
Comment (ian0371) [reply kaiax/gov/headergov/impl/init.go:290]: I didn't fully get what you're saying, but `genesis.Governance` in DB is created differently:
- Private net, `ken init`: `GetGovernanceItemsFromChainConfig` is used, and `SetGenesisGovernance` is not called. It hits this line followed by `genesis.Commit`:
https://github.com/kaiachain/kaia/blob/0412649e00e73edb757ad706a8d93345d6b737b5/blockchain/genesis.go#L209

- Mainnet/Kairos, the very first launch without init: `SetGenesisGovernance` is used, followed by `genesis.Commit`:
https://github.com/kaiachain/kaia/blob/0412649e00e73edb757ad706a8d93345d6b737b5/blockchain/genesis.go#L206

- All networks, launch after init: `SetGenesisGovernance` is called, but *not* committed to DB. Thus, `genesis.Governance` in memory contains the initial thirteen parameters regardless of the genesis in DB, but seems not to be used.
https://github.com/kaiachain/kaia/blob/0412649e00e73edb757ad706a8d93345d6b737b5/blockchain/genesis.go#L242

Comment (blukat29) [reply kaiax/gov/headergov/impl/init.go:290]: It is very unlikely corner case where `genesis.Config.Governance = nil`
- In SetGenesisGovernance, result includes governance.unitprice
- In GetGovernanceItemsFromChainConfig and getGenesisParamNames, result does not include governance.unitprice.
But no use case has `genesis.Config.Governance = nil` though.

File: kaiax/gov/headergov/impl/init.go (hunk 2 above)
Comment (ian0371) [reply kaiax/gov/headergov/impl/init.go:292]: added 292fff4d

File: kaiax/gov/headergov/impl/init.go (hunk 1 above)
Comment (ian0371) [reply kaiax/gov/headergov/impl/init.go:294]: If `genesis.Config.Governance = nil`, then `SetGenesisGovernance` is never called. Even if it is called, it's going to cause nil dereference panic.

Comment (blukat29) [reply kaiax/gov/headergov/impl/init.go:294]: Oh, you're right.

**Review by hyeonLewis — APPROVED** [review]

**Pull request merged**